### PR TITLE
[serve] Add `close` method to `ASGIMessageQueue` to avoid `receive_asgi_messages` hanging

### LINKED ISSUE #38368: [serve] Possible leak of `receive_asgi_messages` tasks on the HTTP proxy
In an Aviary load test, during a burst of load where many requests timed out, it was observed in the metrics that the number of `HTTPProxyActor.receive_asgi_messages` tasks ongoing did not decrease after the load subsided.

image

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

See the issue linked below for observed behavior.

I haven't been able to reproduce this, but think I have determined the case where it can happen:

- In `receive_asgi_messages`, we first get the queue from `self.asgi_receive_queues` then call `await queue.wait_for_message()`.
- The assumption is that `receive_asgi_messages` will always return an `http.disconnect` message to finalize the request, but in cases where there is a timeout on the HTTP proxy, we cancel the `proxy_asgi_receive` task which prevents it from placing that message on the queue.
- When this happens, the queue is deleted from `self.asgi_receive_queues`. However, the task may have already fetched the queue from the map and called `await queue.wait_for_message()`.
- Because the `receive_asgi_messages` cannot be canceled, it would then hang waiting for a message that never arrives, effectively leaking a long-running task to the proxy.

This PR fixes the behavior by adding a `close` method on the queue that's called when the task is cancelled. After `close` is called, all subsequent calls will return immediately so we cannot have a task hang in this manner.

I've added a regression test to `test_request_timeout.py` that fails without this fix and passes with it.

## Related issue number

Closes https://github.com/ray-project/ray/issues/38368

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by shrekris-anyscale — APPROVED** [review]

**Comment by sihanwang41** [comment]
Hi @edoakes , is it possible `receive_asgi_messages` being called before queue is closed? so that `receive_asgi_messages ` task still can't be fulfilled.

**Review by GeneDer — APPROVED** [review]
:shipit:

**Comment by edoakes** [comment]
> Hi @edoakes , is it possible `receive_asgi_messages` being called before queue is closed? so that `receive_asgi_messages ` task still can't be fulfilled.

I'm not totally sure what you mean by this. `receive_asgi_messages` being called before the queue is closed is the common case, if the queue doesn't it exist it'll error. If the queue exists it'll block on `wait_for_message` (which will unblock when the queue is closed).

**Review by sihanwang41 — APPROVED** [review]
Okay, I missed the `self._new_message_event.set()` in the change :)

File: python/ray/serve/tests/test_http_util.py [hunk 1]
```diff
@@ -61,8 +61,11 @@ async def test_asgi_message_queue():
     assert len(list(queue.get_messages_nowait())) == 1
 
     # Check that once the queue is closed, new messages should be rejected and
-    # subsequent calls to wait for messages should return immediately.
+    # ongoing and subsequent calls to wait for messages should return immediately.
```
Comment (sihanwang41) [inline python/ray/serve/tests/test_http_util.py:64]: nice 👍

**Pull request merged**